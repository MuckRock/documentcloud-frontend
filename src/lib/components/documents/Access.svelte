<script lang="ts">
  import { _ } from "svelte-i18n";
  import { Globe24, Lock24, Organization24 } from "svelte-octicons";

  import type { Document, Access } from "@/lib/api/types";
  import { isOrg } from "@/lib/api/accounts";
  import Avatar from "$lib/components/accounts/Avatar.svelte";
  import Flex from "$lib/components/common/Flex.svelte";
  import Metadata from "$lib/components/common/Metadata.svelte";
  import type { Level } from "$lib/components/inputs/AccessLevel.svelte";
  import SidebarItem from "$lib/components/sidebar/SidebarItem.svelte";

  export let document: Document;

  export const levels: Level[] = [
    {
      value: $_("access.private.value") as Access,
      title: $_("access.private.title"),
      description: $_("access.private.description"),
      icon: Lock24,
    },
    {
      value: $_("access.organization.value") as Access,
      title: $_("access.organization.title"),
      description: $_("access.organization.description"),
      icon: Organization24,
    },
    {
      value: $_("access.public.value") as Access,
      title: $_("access.public.title"),
      description: $_("access.public.description"),
      icon: Globe24,
    },
  ];

  $: access = levels.find((level) => level.value === document.access);
</script>

<SidebarItem
  inline
  title={access.description}
  --color="var(--gray-5)"
  --font-size="var(--font-lg)"
>
  <svelte:component this={access.icon} slot="start" />
  {access.title}
</SidebarItem>
{#if document.access === "organization" && isOrg(document.organization)}
  <Metadata key={$_("sidebar.sharedWith")}>
    <Flex>
      <Avatar org={document.organization} />
      {document.organization.name}
    </Flex>
  </Metadata>
{/if}
