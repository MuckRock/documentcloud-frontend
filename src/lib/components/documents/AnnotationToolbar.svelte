<script lang="ts">
  import { getContext } from "svelte";
  import type { Writable } from "svelte/store";
  import { _ } from "svelte-i18n";
  import { Comment16, Question16 } from "svelte-octicons";
  import type { ViewerMode } from "$lib/api/types";
  import Flex from "$lib/components/common/Flex.svelte";
  import PageToolbar from "$lib/components/common/PageToolbar.svelte";
  import Tooltip from "@/common/Tooltip.svelte";
  import Button from "$lib/components/common/Button.svelte";

  const currentMode: Writable<ViewerMode> = getContext("currentMode");
</script>

<PageToolbar>
  <Flex align="center" slot="left">
    <h3 class="title">
      <Comment16 />
      {$_("annotate.title")}
    </h3>
    <Tooltip caption={$_("annotate.instructions")}>
      <Question16 fill="var(--blue-3)" />
    </Tooltip>
  </Flex>
  <Button
    size="small"
    slot="right"
    mode="primary"
    on:click={() => ($currentMode = "document")}
  >
    Done
  </Button>
</PageToolbar>

<style>
  .title {
    font-weight: var(--font-semibold);
    font-size: var(--font-md);
    display: inline-flex;
    align-items: center;
    gap: 0.5rem;
    margin-left: 0.5rem;
  }
</style>
