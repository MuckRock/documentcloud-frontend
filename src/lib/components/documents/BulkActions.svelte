<!-- @component
This is a menu that wraps all logic around bulk actions for documents.
Most actual actions are deferred to their own forms, so this is more of a switchboard.
-->
<script context="module" lang="ts">
  type Action = "share" | "edit" | "data" | "reprocess" | "delete" | "project";
  type ActionDetail = [label: string, icon: ComponentType];
</script>

<script lang="ts">
  import type { Readable } from "svelte/store";
  import type { Document, Maybe, Nullable } from "$lib/api/types";

  import { getContext, type ComponentType } from "svelte";
  import { _ } from "svelte-i18n";
  import {
    Alert16,
    FileDirectory16,
    IssueReopened16,
    Pencil16,
    Share16,
    Tag16,
  } from "svelte-octicons";

  import Modal from "../layouts/Modal.svelte";
  import Portal from "../layouts/Portal.svelte";
  import SidebarItem from "../sidebar/SidebarItem.svelte";

  // forms
  import ConfirmDelete from "../forms/ConfirmDelete.svelte";
  import EditDataMany from "../forms/EditDataMany.svelte";
  import EditMany from "../forms/EditMany.svelte";
  import Reprocess from "../forms/Reprocess.svelte";
  import Projects from "../forms/Projects.svelte";

  import { getCurrentUser } from "$lib/utils/permissions";
  import Share from "./Share.svelte";

  export let afterClick: Maybe<() => void> = undefined;

  const me = getCurrentUser();
  const selected: Readable<Document[]> = getContext("selected");

  const actions: Record<Action, ActionDetail> = {
    share: [$_("bulk.actions.share"), Share16],
    edit: [$_("bulk.actions.edit"), Pencil16],
    data: [$_("bulk.actions.data"), Tag16],
    project: [$_("bulk.actions.project"), FileDirectory16],
    reprocess: [$_("bulk.actions.reprocess"), IssueReopened16],
    delete: [$_("bulk.actions.delete"), Alert16],
  };

  let visible: Nullable<Action> = null;

  // svelte 5 will let us do type coercion in templates
  $: toShare = $selected[0]!;
  function show(action: string) {
    visible = action as Action;
  }

  function close() {
    visible = null;
  }
</script>

<!-- for now, we can only share individual documents -->
{#each Object.entries(actions) as [action, [label, icon]]}
  <SidebarItem
    hover
    disabled={action === "share"
      ? $selected?.length !== 1
      : !$me || !$selected?.length}
    on:click={() => {
      show(action);
      afterClick?.();
    }}
  >
    <svelte:component this={icon} slot="start" />
    {label}
  </SidebarItem>
{/each}

{#if visible}
  <Portal>
    <Modal on:close={close}>
      <h1 slot="title">{actions[visible][0]}</h1>

      {#if visible === "share"}
        <Share document={toShare} />
      {/if}

      {#if visible === "edit"}
        <EditMany documents={$selected} on:close={close}>
          {#if $selected?.length}
            <p>{$_("edit.many", { values: { n: $selected?.length } })}</p>
          {/if}
        </EditMany>
      {/if}

      {#if visible === "delete"}
        <ConfirmDelete documents={$selected} on:close={close} />
      {/if}

      {#if visible === "reprocess"}
        <Reprocess documents={$selected} on:close={close} />
      {/if}

      {#if visible === "data"}
        <EditDataMany documents={$selected} on:close={close} />
      {/if}

      {#if visible === "project"}
        <Projects documents={$selected} on:close={close} />
      {/if}
    </Modal>
  </Portal>
{/if}
