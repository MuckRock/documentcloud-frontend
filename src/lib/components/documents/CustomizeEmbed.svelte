<script lang="ts" context="module">
  import { writable } from "svelte/store";
  import { settings, settingsConfig } from "$lib/utils/embed";

  export const embedSettings = writable(settings);
</script>

<script lang="ts">
  import { onMount } from "svelte";
  import { _ } from "svelte-i18n";

  import Flex from "../common/Flex.svelte";
  import Field from "../common/Field.svelte";
  import NumberInput from "../inputs/Number.svelte";

  import { StorageManager } from "$lib/utils/storage";

  // The embed customization allows a user to control settings
  // that customize the presentation of a document embed. The user's
  // preferences should be saved to localStorage for future embeds.
  export let storageManager = new StorageManager("vieweroptions");

  // initialize settings with loaded values
  onMount(() => {
    Object.keys(settingsConfig).forEach((key) => {
      const config = settingsConfig[key];
      const storageValue = storageManager.get(
        String(config.storageIndex),
        config.defaultValue,
      );
      if ($embedSettings[key] === null) {
        $embedSettings[key] = storageValue;
      }
    });
  });
</script>

<div>
  {#each Object.entries(settingsConfig) as [name, config]}
    {#if config.field.type === "hidden"}
      <input type="hidden" {name} value={config.defaultValue} />
    {:else}
      <fieldset>
        <legend>{$_(config.field.label)}</legend>
        {#if config.field.type === "dimension"}
          {#if config.field.automatic}
            <Flex>
              <div class="flexItem">
                <Field inline>
                  <label>
                    <input
                      type="radio"
                      {name}
                      value={config.field.automatic.value}
                      bind:group={$embedSettings[name]}
                    />
                    {$_(config.field.automatic.label)}
                    {#if config.defaultValue === config.field.automatic.value}
                      <span class="default">{$_("share.default")}</span>
                    {/if}
                  </label>
                  <p slot="help">
                    {$_(config.field.automatic.help)}
                  </p>
                </Field>
              </div>

              <div class="flexItem">
                <Field inline>
                  <label>
                    <input
                      type="radio"
                      {name}
                      value={$embedSettings[name] >= 1
                        ? $embedSettings[name]
                        : config.field.fixed.value}
                      bind:group={$embedSettings[name]}
                    />
                    {$_(config.field.fixed.label)}

                    {#if config.defaultValue === config.field.fixed.value}
                      <span class="default">{$_("share.default")}</span>
                    {/if}
                  </label>

                  <p slot="help">
                    {$_(config.field.fixed.help)}
                  </p>
                </Field>
              </div>
            </Flex>
          {/if}
          {#if $embedSettings[name] >= 1}
            <div class="number-input">
              <NumberInput min={1} {name} bind:value={$embedSettings[name]} />
            </div>
          {/if}
        {:else}
          <Flex>
            {#each config.field.options as option}
              <div class="flexItem">
                <Field inline>
                  <label>
                    <input
                      type="radio"
                      {name}
                      value={option.value}
                      bind:group={$embedSettings[name]}
                    />
                    {$_(option.label)}
                    {#if config.defaultValue === option.value}
                      <span class="default">{$_("share.default")}</span>
                    {/if}
                  </label>
                  <p slot="help">{$_(option.help)}</p>
                </Field>
              </div>
            {/each}
          </Flex>
        {/if}
      </fieldset>
    {/if}
  {/each}
</div>

<style>
  fieldset {
    margin: 0.5em 0;
    border: 1px solid var(--gray-2);
    border-radius: 0.25rem;
  }
  .flexItem {
    flex: 0 1 12rem;
    padding: 0 0.5rem;
  }
  .number-input {
    margin-top: 0.5rem;
  }
  .default {
    font-size: var(--font-xs);
    color: var(--gray-4);
    font-weight: var(--font-semibold);
    margin-left: 0.25rem;
  }
  legend {
    color: var(--gray-5);
    font-weight: var(--font-semibold);
    font-size: var(--font-sm);
    padding: 0.5em;
  }
  input[type="radio"] {
    margin: 0 0.25em 0 0;
    transform: translateY(1px);
  }
  label {
    font-weight: var(--font-semibold);
    color: var(--gray-5);
    padding: 0.25em 0.375em;
    border-radius: 0.25em;
    vertical-align: middle;
  }
  label:hover {
    background: var(--blue-1);
  }
</style>
