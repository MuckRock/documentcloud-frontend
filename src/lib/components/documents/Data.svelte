<script lang="ts">
  import type { Document } from "$lib/api/types";

  import { _ } from "svelte-i18n";
  import { ChevronDown12, Pencil16, Tag16, Tag24 } from "svelte-octicons";

  import Action from "$lib/components/common/Action.svelte";
  import Empty from "$lib/components/common/Empty.svelte";
  import Flex from "$lib/components/common/Flex.svelte";
  import KV from "$lib/components/common/KV.svelte";
  import SidebarItem from "$lib/components/sidebar/SidebarItem.svelte";

  // editing UI
  import EditData from "@/lib/components/forms/EditData.svelte";
  import Modal from "$lib/components/layouts/Modal.svelte";
  import Portal from "$lib/components/layouts/Portal.svelte";

  import * as search from "$lib/utils/search";
  import SidebarGroup from "../sidebar/SidebarGroup.svelte";

  export let document: Document;

  let edit = false;

  $: tags = document.data["_tag"];
  $: data = Object.entries(document.data).filter(([k, v]) => k !== "_tag");
  $: empty = Object.keys(document.data).length === 0;
</script>

<SidebarGroup name="projects:viewer">
  <SidebarItem slot="title">
    <Tag16 />
    {$_("sidebar.data.title")}
  </SidebarItem>
  <div slot="action">
    {#if document.edit_access}
      <Action on:click={() => (edit = true)} icon={Pencil16}>
        {$_("common.edit")}
      </Action>
    {/if}
  </div>
  <div>
    {#if empty}
      <Empty icon={Tag24}>
        <p>{$_("sidebar.data.empty")}</p>
      </Empty>
    {:else}
      <Flex direction="column" gap={1}>
        {#if tags}
          <Flex wrap class="tags">
            {#each tags as tag}
              <KV
                tag
                value={tag}
                href={search.searchUrl(search.tag(tag)).href}
              />
            {/each}
          </Flex>
        {/if}

        {#if data.length}
          <Flex wrap class="data">
            {#each data as [key, values]}
              {#each values as value}
                <KV
                  {key}
                  {value}
                  href={search.searchUrl(search.kv(key, value)).href}
                />
              {/each}
            {/each}
          </Flex>
        {/if}
      </Flex>
    {/if}
  </div>
</SidebarGroup>

{#if edit}
  <Portal>
    <Modal on:close={() => (edit = false)}>
      <h2 slot="title">{$_("data.title")}</h2>
      <EditData {document} on:close={() => (edit = false)} />
    </Modal>
  </Portal>
{/if}
