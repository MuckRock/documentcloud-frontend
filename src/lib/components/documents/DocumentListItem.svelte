<!--
@component
This is a list item showing one document. 
It's deliberately minimal and can be wrapped in other components to add additional functionality.

If we're in an embed, we want to open links to documents in new tabs and hide the access label.
-->
<script lang="ts">
  import type { Access, Document, Project } from "$lib/api/types";

  import DOMPurify from "isomorphic-dompurify";
  import { getContext } from "svelte";
  import { _ } from "svelte-i18n";
  import { Alert24, Hourglass24, File24 } from "svelte-octicons";

  import DocAccess, { getLevel } from "../documents/Access.svelte";
  import Flex from "../common/Flex.svelte";
  import KV from "../common/KV.svelte";
  import NoteTab from "../viewer/NoteTab.svelte";

  import { IMAGE_WIDTHS_MAP } from "@/config/config.js";
  import {
    canonicalUrl,
    pageImageUrl,
    userOrgString,
  } from "@/lib/api/documents";
  import { pageSizesFromSpec } from "@/api/pageSize.js";

  export let document: Document;

  const embed: boolean = getContext("embed");

  const width = IMAGE_WIDTHS_MAP.get("thumbnail");

  // this can be updated later if we want different icons
  const statusIcons = {
    error: Alert24,
    nofile: File24,
    pending: Hourglass24,
  };

  $: sizes = document.page_spec ? pageSizesFromSpec(document.page_spec) : null;
  $: aspect = sizes ? sizes[0] : 11 / 8.5; // fallback to US letter for now
  $: height = width * aspect;
  $: date = new Date(document.created_at).toDateString();
  $: projects = document.projects?.every((p) => typeof p === "object")
    ? (document.projects as Project[])
    : []; // only show projects if we've loaded them

  $: hasPublicNotes = document.notes?.some((note) => note.access === "public");
  $: hasOrgNotes = document.notes?.some(
    (note) => note.access === "organization",
  );
  $: hasPrivateNotes = document.notes?.some(
    (note) => note.access === "private",
  );
  $: level = getLevel(document.access);

  $: tabs = [
    hasPublicNotes && ("public" as Access),
    hasOrgNotes && ("organization" as Access),
    hasPrivateNotes && ("private" as Access),
  ].filter(Boolean);

  function clean(html: string) {
    return DOMPurify.sanitize(html, { ALLOWED_TAGS: [] });
  }

  function onError() {
    document.status = "error";
  }
</script>

<a
  href={canonicalUrl(document).href}
  class="document-list-item"
  target={embed ? "_blank" : undefined}
>
  <div class="thumbnail">
    {#if document.status === "success" || document.status === "readable"}
      <img
        src={pageImageUrl(document, 1, "thumbnail").href}
        alt="Page 1, {document.title}"
        width="{width}px"
        height="{height}px"
        loading="lazy"
        on:error={onError}
      />
    {:else}
      <div class="fallback {document.status}">
        <svelte:component this={statusIcons[document.status]} />
      </div>
    {/if}

    {#if tabs.length > 0}
      <div class="tabs">
        {#each tabs as access}
          <NoteTab {access} size="small" />
        {/each}
      </div>
    {/if}
  </div>

  <div class="documentInfo">
    <h3>{document.title}</h3>
    <p class="meta">
      {$_("documents.pageCount", { values: { n: document.page_count } })} -
      {#if document.notes && document.notes.length > 0}
        {$_("documents.noteCount", { values: { n: document.notes.length } })} -
      {/if}
      {#if userOrgString(document)}
        {userOrgString(document)} -
      {/if}
      {date}
    </p>
    {#if document.description}
      <p class="description">
        {clean(document.description)}
      </p>
    {/if}
    <Flex align="center" gap={0.625} class="actions">
      {#if !embed && level}
        <div style="font-size: var(--font-sm)">
          <DocAccess {level} />
        </div>
      {/if}
      {#each projects as project}
        <KV key="Project" value={project.title} />
      {/each}
    </Flex>
  </div>
</a>

<style>
  .document-list-item {
    flex: 1 0 0;
    display: flex;
    max-width: 100%;
    min-width: 0;
    align-items: center;
    align-self: stretch;
    gap: 0.5rem;
    padding: 0.75rem;
    border-radius: 0.25rem;
    color: inherit;
    text-decoration: none;
    background-color: transparent;
    border: 1px solid transparent;
  }

  .document-list-item:hover,
  .document-list-item:focus {
    background-color: var(--blue-1);
    border-color: var(--blue-2);
  }

  .thumbnail {
    display: flex;
    flex-direction: column;
    align-items: center;
    position: relative;
  }

  .thumbnail img,
  .thumbnail .fallback {
    border-radius: 0.125rem;
    border: 1px solid var(--gray-2, #cbcbcb);
    background: var(--white, #fff);
    box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.1);
  }

  .thumbnail .tabs {
    position: absolute;
    left: -0.5rem;
    top: 0.5rem;
    display: flex;
    flex-direction: column;
    gap: 0.125rem;
  }

  .documentInfo {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    gap: 0.25rem;
    flex: 1 0 0;
    align-self: stretch;
    min-width: 0;
  }

  h3 {
    color: var(--gray-5, #233944);
    font-size: var(--font-md, 1rem);
    font-style: normal;
    font-weight: var(--font-semibold, 600);
    line-height: normal;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    max-width: 100%;
  }

  .meta {
    color: var(--gray-4, #5c717c);
    font-size: 0.75rem;
    font-style: normal;
    font-weight: var(--font-regular, 400);
    line-height: normal;
  }

  .description {
    color: var(--gray-5, #233944);
    font-size: 0.75rem;
    font-style: normal;
    font-weight: var(--font-regular, 400);
    line-height: normal;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    max-width: 100ch;
    width: 100%;
  }

  .fallback {
    display: flex;
    align-items: center;
    justify-content: center;

    background-color: var(--white);
    fill: var(--gray-3);
    /* US letter size, thumbnail width */
    height: calc(60px * 11 / 8.5);
    width: 60px;
  }

  .fallback.error {
    fill: var(--red-3, #f00);
  }
</style>
