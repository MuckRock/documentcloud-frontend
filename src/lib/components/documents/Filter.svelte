<script lang="ts" context="module">
  export interface FilterFields {
    access?: string;
    minPages?: number;
    maxPages?: number;
    projects?: Project[];
    users?: User[];
    orgs?: Org[];
  }

  export const defaultFilters: FilterFields = {
    access: "",
    minPages: undefined,
    maxPages: undefined,
    projects: [],
    users: [],
    orgs: [],
  };
</script>

<script lang="ts">
  import deepEqual from "fast-deep-equal";
  import { _ } from "svelte-i18n";
  import {
    ChevronDown16,
    Eye16,
    FileDirectory16,
    Filter16,
    Hash16,
    LinkExternal16,
    Organization16,
    Person16,
  } from "svelte-octicons";
  import Dropdown from "../common/Dropdown.svelte";
  import Menu from "../common/Menu.svelte";
  import SidebarItem from "../sidebar/SidebarItem.svelte";
  import type { Org, Project, User } from "@/lib/api/types";
  import Button from "../common/Button.svelte";
  import { list as listProjects } from "@/lib/api/projects";
  import { listUsers, listOrgs } from "@/lib/api/accounts";
  import Select from "../inputs/Select.svelte";

  export let filters: FilterFields = defaultFilters;

  async function loadProjects(text: string): Promise<Project[]> {
    const { data, error } = await listProjects({ query: text });
    if (data && !error) {
      return data.results;
    } else if (!data) {
      console.warn("Missing response data");
      return [];
    } else {
      throw new Error(error?.message);
    }
  }

  async function loadUsers(text: string): Promise<User[]> {
    const { data, error } = await listUsers({ name__istartswith: text });
    if (data && !error) {
      return data.results;
    } else if (!data) {
      console.warn("Missing response data");
      return [];
    } else {
      throw new Error(error?.message);
    }
  }

  async function loadOrgs(text: string): Promise<Org[]> {
    const { data, error } = await listOrgs({ name__istartswith: text });
    if (data && !error) {
      return data.results;
    } else if (!data) {
      console.warn("Missing response data");
      return [];
    } else {
      throw new Error(error?.message);
    }
  }
</script>

<Dropdown>
  <SidebarItem slot="anchor">
    <Filter16 slot="start" />
    {$_("documentBrowser.filter.label")}
    <ChevronDown16 slot="end" />
  </SidebarItem>
  <Menu --overflow-y="visible">
    <div class="filters">
      <Select
        items={[
          {
            label: $_("documentBrowser.filter.access.private"),
            value: "private",
          },
          {
            label: $_("documentBrowser.filter.access.public"),
            value: "public",
          },
        ]}
        bind:value={filters.access}
        name="access"
        placeholder="Access"
        clearable
      >
        <Eye16 slot="prepend" />
      </Select>
      <Select
        loadOptions={loadUsers}
        bind:value={filters.users}
        name="users"
        placeholder={$_("documentBrowser.filter.users")}
        itemId="id"
        label="name"
        multiple
      >
        <Person16 slot="prepend" />
      </Select>
      <Select
        loadOptions={loadOrgs}
        bind:value={filters.orgs}
        name="orgs"
        placeholder={$_("documentBrowser.filter.orgs")}
        itemId="id"
        label="name"
        multiple
      >
        <Organization16 slot="prepend" />
      </Select>
      <Select
        loadOptions={loadProjects}
        bind:value={filters.projects}
        name="projects"
        placeholder={$_("documentBrowser.filter.projects")}
        itemId="id"
        label="title"
        multiple
      >
        <FileDirectory16 slot="prepend" />
      </Select>
    </div>
    <fieldset class="pageCounts">
      <label class="min pages">
        <Hash16 />
        <input
          type="number"
          min="1"
          placeholder={$_("documentBrowser.filter.minPages")}
          bind:value={filters.minPages}
        />
      </label>
      <label class="max pages">
        <Hash16 />
        <input
          type="number"
          min="1"
          placeholder={$_("documentBrowser.filter.maxPages")}
          bind:value={filters.maxPages}
        />
      </label>
    </fieldset>
    <Button ghost mode="primary">
      <LinkExternal16 />
      {$_("documentBrowser.filter.advancedHelp")}
    </Button>
  </Menu>
</Dropdown>

<style>
  :global(.filters .select) {
    font-size: var(--font-sm);
  }
  .filters {
    display: flex;
    flex-direction: column;
    gap: 0.5rem;
    padding: 0.25rem;
  }
  .pageCounts {
    margin: 0;
    padding: 0;
    border: none;
    display: flex;
    align-items: center;
    gap: 0.5rem;
    padding: 0 0.25rem 0.25rem;
  }
  .pages {
    flex: 1 1 8rem;
    display: flex;
    align-items: center;
    gap: 0.25rem;
    color: var(--gray-5);
    fill: var(--gray-4);
    padding: 0.25rem;
    border-radius: 0.5rem;
    border: 1px solid var(--gray-2);
    &:hover {
      border-color: var(--gray-3);
    }
    &:active,
    &:focus,
    &:focus-within {
      border-color: var(--blue-3);
      outline: 2px solid var(--blue-3);
    }
    & input {
      flex: 1 1 auto;
      appearance: none;
      border: none;
      background: transparent;
      font-family: var(--font-sans);
      font-size: var(--font-md);
      &:focus {
        outline: none;
      }
    }
  }
</style>
