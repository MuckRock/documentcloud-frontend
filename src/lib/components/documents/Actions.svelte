<script lang="ts">
  import { _ } from "svelte-i18n";
  import {
    Download16,
    History16,
    IssueReopened16,
    Pencil16,
    Share16,
    Trash16,
  } from "svelte-octicons";

  import { canonicalUrl, pdfUrl } from "$lib/api/documents";
  import type { Document, Org, User } from "$lib/api/types";
  import Button from "$lib/components/common/Button.svelte";
  import Reprocess from "$lib/components/forms/Reprocess.svelte";
  import ConfirmDelete from "$lib/components/forms/ConfirmDelete.svelte";
  import Edit from "$lib/components/forms/Edit.svelte";
  import Portal from "$lib/components/layouts/Portal.svelte";
  import Modal from "$lib/components/layouts/Modal.svelte";

  import Share from "./Share.svelte";
  import PremiumBadge from "@/premium-credits/PremiumBadge.svelte";
  import Premium from "@/common/icons/Premium.svelte";
  import RevisionControl from "../forms/RevisionControl.svelte";
  import Revisions from "./Revisions.svelte";
  import UpgradePrompt from "@/premium-credits/UpgradePrompt.svelte";
  import { getUpgradeUrl } from "$lib/api/accounts";

  export let document: Document;
  export let user: User;

  let shareOpen = false;
  let editOpen = false;
  let revisionsOpen = false;
  let reprocessOpen = false;
  let deleteOpen = false;

  function relativeUrl(document: Document, path: string) {
    return new URL(path, canonicalUrl(document)).href;
  }

  $: canEdit = document.edit_access;
  $: organization =
    typeof user?.organization === "object" ? (user.organization as Org) : null;
  $: plan = organization?.plan ?? "Free";
</script>

<div class="actions wideGap">
  {#if canEdit}
    <div class="actions">
      <Button
        ghost
        mode="primary"
        minW={false}
        on:click={() => (editOpen = true)}
      >
        <Pencil16 />
        {$_("sidebar.edit")}
      </Button>
      <Button
        ghost
        mode="premium"
        minW={false}
        on:click={() => (revisionsOpen = true)}
      >
        <History16 />
        {$_("sidebar.revisions")}
        <Premium />
      </Button>
    </div>
  {/if}
  <div class="actions">
    <Button ghost on:click={() => (shareOpen = true)}>
      <Share16 />
      {$_("sidebar.shareEmbed")}
    </Button>
    <Button ghost href={pdfUrl(document).toString()} download target="_blank">
      <Download16 />
      {$_("sidebar.download")}
    </Button>
  </div>
  {#if canEdit}
    <div class="actions">
      <Button
        ghost
        mode="danger"
        disabled={document.status === "nofile"}
        on:click={() => (reprocessOpen = true)}
      >
        <IssueReopened16 />
        {$_("sidebar.reprocess")}
      </Button>
      <Button ghost mode="danger" on:click={() => (deleteOpen = true)}>
        <Trash16 />
        {$_("sidebar.delete")}
      </Button>
    </div>
  {/if}
</div>

{#if canEdit && editOpen}
  <Portal>
    <Modal on:close={() => (editOpen = false)}>
      <h1 slot="title">{$_("edit.title")}</h1>
      <Edit {document} on:close={() => (editOpen = false)} />
    </Modal>
  </Portal>
{/if}
{#if revisionsOpen}
  <Portal>
    <Modal on:close={() => (revisionsOpen = false)}>
      <h1 slot="title">
        {$_("dialogRevisionsDialog.heading")}
        <PremiumBadge />
      </h1>
      <div>
        {#if plan !== "Free"}
          <RevisionControl {document} />
          <hr class="divider" />
          <Revisions {document} />
        {:else}
          <UpgradePrompt
            message={$_("dialogRevisionsDialog.upgrade.message")}
            callToAction={$_("dialogRevisionsDialog.upgrade.adminCta")}
            href={getUpgradeUrl(organization).href}
          />
        {/if}
      </div>
    </Modal>
  </Portal>
{/if}
{#if shareOpen}
  <Portal>
    <Modal on:close={() => (shareOpen = false)}>
      <h1 slot="title">{$_("sidebar.shareEmbed")}</h1>
      <Share {document} />
    </Modal>
  </Portal>
{/if}
{#if reprocessOpen}
  <Portal>
    <Modal on:close={() => (reprocessOpen = false)}>
      <h1 slot="title">{$_("dialogReprocessDialog.title")}</h1>
      <Reprocess
        documents={[document]}
        on:close={() => (reprocessOpen = false)}
      />
    </Modal>
  </Portal>
{/if}

{#if deleteOpen}
  <Portal>
    <Modal on:close={() => (deleteOpen = false)}>
      <h1 slot="title">{$_("delete.title")}</h1>
      <ConfirmDelete
        documents={[document]}
        on:close={() => (deleteOpen = false)}
      />
    </Modal>
  </Portal>
{/if}

<style>
  .actions {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    gap: 0.5rem;
  }
  .wideGap {
    gap: 1rem;
  }
</style>
