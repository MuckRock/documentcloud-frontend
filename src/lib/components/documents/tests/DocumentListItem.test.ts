import { vi, describe, it, beforeEach, expect } from "vitest";
import { render } from "@testing-library/svelte";
import { document } from "@/test/fixtures/documents";
import DocumentListItem from "../DocumentListItem.svelte";

describe("DocumentListItem", async () => {
  beforeEach(() => {
    // Mock Date.now() to return a fixed timestamp
    vi.spyOn(Date, "now").mockReturnValue(1620000000000); // Mocked timestamp
  });
  it("renders", () => {
    let result = render(DocumentListItem, { document });
    expect(result.container).toMatchSnapshot();
  });
});
