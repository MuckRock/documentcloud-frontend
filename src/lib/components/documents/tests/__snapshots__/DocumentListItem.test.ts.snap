// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DocumentListItem > renders 1`] = `
<div>
  <a
    class="document-list-item svelte-o1ttsh"
    href="https://www.dev.documentcloud.org/documents/24002098-quarterly-reports-created-pursuant-to-the-letter-from-the-national-archives-and-records-administration-dated-june-1-2018/"
  >
    <div
      class="thumbnail svelte-o1ttsh"
    >
      <img
        alt="Page 1, Quarterly Reports created pursuant to the letter from the National Archives and Records Administration dated June 1, 2018"
        class="svelte-o1ttsh"
        height="77.64705882352942px"
        loading="lazy"
        src="https://s3.documentcloud.org/documents/24002098/pages/quarterly-reports-created-pursuant-to-the-letter-from-the-national-archives-and-records-administration-dated-june-1-2018-p1-thumbnail.gif"
        width="60px"
      />
       
    </div>
     
    <div
      class="documentInfo svelte-o1ttsh"
    >
      <div
        class="head svelte-o1ttsh"
      >
        <h3
          class="title svelte-o1ttsh"
        >
          Quarterly Reports created pursuant to the letter from the National Archives and Records Administration dated June 1, 2018
        </h3>
         
        <div
          class="access svelte-o1ttsh"
        >
          <div
            class="access public svelte-o63adk"
          >
            <svg
              class="octicon octicon-globe"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 1c6.075 0 11 4.925 11 11s-4.925 11-11 11S1 18.075 1 12 5.925 1 12 1Zm3.241 10.5v-.001c-.1-2.708-.992-4.904-1.89-6.452a13.919 13.919 0 0 0-1.304-1.88L12 3.11l-.047.059c-.354.425-.828 1.06-1.304 1.88-.898 1.547-1.79 3.743-1.89 6.451Zm-12.728 0h4.745c.1-3.037 1.1-5.49 2.093-7.204.39-.672.78-1.233 1.119-1.673C6.11 3.329 2.746 7 2.513 11.5Zm18.974 0C21.254 7 17.89 3.329 13.53 2.623c.339.44.729 1.001 1.119 1.673.993 1.714 1.993 4.167 2.093 7.204ZM8.787 13c.182 2.478 1.02 4.5 1.862 5.953.382.661.818 1.29 1.304 1.88l.047.057.047-.059c.354-.425.828-1.06 1.304-1.88.842-1.451 1.679-3.471 1.862-5.951Zm-1.504 0H2.552a9.505 9.505 0 0 0 7.918 8.377 15.773 15.773 0 0 1-1.119-1.673C8.413 18.085 7.47 15.807 7.283 13Zm9.434 0c-.186 2.807-1.13 5.085-2.068 6.704-.39.672-.78 1.233-1.118 1.673A9.506 9.506 0 0 0 21.447 13Z"
              />
            </svg>
             
            Public
          </div>
        </div>
      </div>
       
      <p
        class="meta svelte-o1ttsh"
      >
        20 pages
         -
      
         
        Mitchell Kotler (MuckRock)
         -
         
        Mon Oct 02 2023
      </p>
       
      <div
        class="data svelte-o1ttsh"
      />
    </div>
  </a>
</div>
`;
